### PR TITLE
fix getCaret selection

### PR DESCRIPTION
**Pull request by iamJoeTaylor**
cc @square/field-kit-contributors

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,21 @@
...
+    "credit-card",
+    "credit card",
+    "ssn",
+    "social",
+    "expiration",
+    "date",
+    "field",
+    "fieldkit",
+    "field-kit"
+  ],
```
Comment (eventualbuddha) [inline package.json]: This is unrelated.

File: src/caret.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ if (!document) {
   };
 } else if(document.selection) {
   getCaret = element => {
+    const selection = document.selection;
```
Comment (eventualbuddha) [inline src/caret.js]: Can we write a test that fails without this?

File: package.json (hunk 1 above)
Comment (iamJoeTaylor) [inline package.json]: Fixed

File: src/caret.js (hunk 2 above)
Comment (iamJoeTaylor) [inline src/caret.js]: Done

File: src/caret.js [hunk 3]
```diff
@@ -1,43 +1,46 @@
...
-  };
-} else if(document.selection) {
-  getCaret = element => {
-    const value = element.value;
-    let range = selection.createRange().duplicate();
+    setCaret = (element, start, end) => {
+      element.selectionStart = start;
+      element.selectionEnd = end;
+    };
+  } else if(_document.selection) {
```
Comment (eventualbuddha) [inline src/caret.js]: nit: space between `if` and `(`

Comment (iamJoeTaylor) [inline src/caret.js]: fixed

File: src/caret.js [hunk 4]
```diff
@@ -1,43 +1,46 @@
-let getCaret;
-let setCaret;
+export default function installCaret(_document = document) {
```
Comment (eventualbuddha) [inline src/caret.js]: Is this called with an argument anywhere? I can't find anything in this diff.

Comment (iamJoeTaylor) [inline src/caret.js]: Good catch. I had a file waiting to be committed, the tests no less. So yes the tests do use this option.

File: test/unit/caret_test.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      })).to.throw(/Caret unknown input selection capabilities/);
+    });
+  });
+
+  describe('with selectionStart', () => {
+    let getCaret, setCaret;
+
+    beforeEach(() => {
+      const caret = installCaret({
+        createElement: () => { return { selectionStart: true }; }
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: `() => ({ selectionStart: true })`

File: test/unit/caret_test.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        createElement: () => { return {}; }
+      })).to.throw(/Caret unknown input selection capabilities/);
+    });
+  });
+
+  describe('with selectionStart', () => {
+    let getCaret, setCaret;
+
+    beforeEach(() => {
+      const caret = installCaret({
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: `({ getCaret, setCaret }) = …`

File: test/unit/caret_test.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  describe('with no document', () => {
+    it('throws an error', () => {
+      expect(installCaret.bind(this, null)).to.throw(/Caret does not have access to document/);
+    });
+  });
+
+  describe('with no native caret', () => {
+    it('throws an error', () => {
+      expect(installCaret.bind(this, {
+        createElement: () => { return {}; }
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: `() => ({})`

File: test/unit/caret_test.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
+import sinon from 'sinon';
+import {expect} from 'chai';
+import installCaret from '../../src/caret';
+
+describe('Caret', () => {
+  describe('with no document', () => {
+    it('throws an error', () => {
+      expect(installCaret.bind(this, null)).to.throw(/Caret does not have access to document/);
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: I wonder if `expect(() => installCaret(null)).to…` is more readable?

File: test/unit/caret_test.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      });
+    });
+  });
+
+  describe('with document.selection', function() {
+    let getCaret, setCaret;
+
+    beforeEach(() => {
+      const caret = installCaret({
+        createElement: () => { return {}; },
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: `() => ({})`

File: test/unit/caret_test.js [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        selection: {
+          createRange: () => {
+            return {
+              duplicate: () => {
+                const selectedText = this.selectedText;
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: The context of these `this.selectedText` gets confusing in this section. In this one it is referring to the `this` inside the callback for `describe`. Inside the `TextRange` class it's an instance of `TextRange`. Can we move `TextRange` to a support file or the top of this one?

File: test/unit/caret_test.js [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+                  get text() {
+                    return this.selectedText;
+                  }
+
+                  moveEnd() {
+                    this.selectedText = selectedText;
+                  }
+
+                  moveStart(noop, negLength) {
+                    if(selectedText) {
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: nit: space between `if` and `(`

File: test/unit/caret_test.js [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+                      this.selectedText = selectedText;
+                    } else {
+                      this.selectedText = '';
+                      for(let i = 0; i < -negLength; i++) {
```
Comment (eventualbuddha) [inline test/unit/caret_test.js]: nit: space between `for` and `(`

File: test/unit/caret_test.js (hunk 8 above)
Comment (iamJoeTaylor) [inline test/unit/caret_test.js]: It is, but I was having issue with the error throwing. This is the only way it would actually catch. Seems stupid to me... but IDK.

File: test/unit/caret_test.js (hunk 9 above)
Comment (iamJoeTaylor) [inline test/unit/caret_test.js]: Thanks, I wasn't sure how to do this.

File: test/unit/caret_test.js (hunk 8 above)
Comment (eventualbuddha) [inline test/unit/caret_test.js]: I think this is fixed in https://github.com/square/field-kit/pull/48.

Comment (iamJoeTaylor) [inline test/unit/caret_test.js]: :+1:

**Comment by eventualbuddha** [comment]
:+1:
